### ISSUE TITLE
New principle: CSS and HTML syntax should be side effect free

### ISSUE THREAD

**Issue by LeaVerou** [w3ctag/design-principles#336]
This was brought up in our [design review of `contain-intrinsic-size: auto`](https://github.com/w3ctag/design-reviews/issues/624).

While the design principle of CSS we were referring to seems obvious, it's still nowhere to be found in our design principles.

**Comment by annevk** [comment]
Note that depending on how you phrase this, it would outlaw `<script>`, `<style>`, `<iframe>`, `background-image`, `cursor`, etc.

**Comment by LeaVerou** [comment]
@annevk This was brought up in one of our breakouts this week, and we are not sure we understand how it would outlaw `<style>`, `background-image`, or `cursor`, could you please elaborate?
And would it outlaw `<iframe>` because it can contain scripts that access the parent, or a different reason?

**Comment by annevk** [comment]
All those features have side effects in that they can cause network fetches, or cause CSS to be applied to a document. (And yeah, `iframe` is essentially a more powerful version of `script`; it ends up creating its own execution environment.)

**Comment by LeaVerou** [comment]
Yes, but the CSS applied to the document only as long as the CSS code is actually present in `<style>`, not after, therefore it is not a side effect.
Same for any network fetches, minus any race conditions where e.g. the element that triggered the fetch gets removed right after the fetch was triggered. Am I missing something?

**Comment by annevk** [comment]
So it's not a side effect in your model if insert + remove is essentially a no-op?

I'm not sure how network fetches are similar? They seem like very much a side effect (and can cause their own side effects, such as setting cookies).

(`style` can also block subsequent `script` elements.)

**Comment by LeaVerou** [comment]
Yes, the design review that sparked this proposed a feature for CSS which would have violated the current implicit rule that rendered style is a function of the current state (and not past states that have stopped applying). 

I do see your point however, and we should be careful in how this is phrased to allow for any side effects from network requests, otherwise anything sending network requests would be off-limits! Might even need to scope it to CSS explicitly, not sure.
